src/actions/receive-profile.js

```diff
@@ -322,6 +328,187 @@ export function finalizeFullProfileView(
   };
 }
 
+/**
+ * This is a small utility to extract the origin from a URL, to build the origins-based
+ * profile view.
+ */
+function getOrigin(urlString: string): string {
+  if (urlString.startsWith('chrome://')) {
+    return urlString;
+  }
+  try {
+    const url = new URL(urlString);
+    if (url.origin === 'null') {
+      // This can happen for "about:newtab"
+      return urlString;
+    }
+    return url.origin;
+  } catch {
+    // This failed, maybe it's an internal URL.
+    return urlString;
+  }
+}
+
+/**
+ * Finalize the profile state for the origin-based view. This is an experimental
+ * view for fission. It's not turned on by default. Note, that this function
+ * probably needs a lot of work to become more correct to handle everything,
+ * so it shouldn't be trusted too much at this time.
+ */
+export function finalizeOriginProfileView(
```

Comment:
Not for this PR but I feel like we can make use of active-tab's code in most of the places and de-duplicate the code a lot. But it's fine by me for now since this is an experiment. We can do it later.

Reply:
Interesting, I'm not sure which code can be shared here. One thing that's nice about this new architecture is that it gives a mechanism to switch on how we compute the initial views.